state/factory/statedb.go

```diff
@@ -232,11 +250,28 @@ func (sdb *stateDB) NewBlockBuilder(
 	ap actpool.ActPool,
 	sign func(action.Envelope) (*action.SealedEnvelope, error),
 ) (*block.Builder, error) {
+	bcCtx := protocol.MustGetBlockchainCtx(ctx)
+	expectedBlockHeight := bcCtx.Tip.Height + 1
 	ctx = protocol.WithRegistry(ctx, sdb.registry)
+	var (
+		ws  *workingSet
+		err error
+	)
 	sdb.mutex.RLock()
 	currHeight := sdb.currentChainHeight
 	sdb.mutex.RUnlock()
-	ws, err := sdb.newWorkingSet(ctx, currHeight+1)
+	switch {
+	case currHeight+1 < expectedBlockHeight:
+		parent, ok := sdb.workingsets.Get(bcCtx.Tip.Hash)
```

Comment:
the key in `workingsets` is not blockHash, we may need cache ws by hash meanwhile